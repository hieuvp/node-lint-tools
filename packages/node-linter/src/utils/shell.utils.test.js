const { run } = require('runjs');
const { exec } = require('./shell.utils');

jest.mock('runjs', () => ({ run: jest.fn() }));

describe('exec', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('command', () => {
    beforeEach(() => {
      run.mockResolvedValue();
    });

    [[], true, undefined, null, 1].forEach(command => {
      it(`should throw an error because "${JSON.stringify(command)}" is not a string`, () => {
        const fn = () => exec(command);
        expect(fn).toThrowErrorMatchingSnapshot();
      });
    });

    ['', ' ', '     '].forEach(command => {
      it(`should throw an error because "${command}" is an empty string`, () => {
        const fn = () => exec(command);
        expect(fn).toThrowErrorMatchingSnapshot();
      });
    });

    it('should recompose a concise command without leading spaces', async () => {
      const command = '  ls -lia';
      await exec(command);
      expect(run.mock.calls[0][0]).toEqual('ls -lia');
    });

    it('should recompose a concise command without trailing spaces', async () => {
      const command = 'ls -lia   ';
      await exec(command);
      expect(run.mock.calls[0][0]).toEqual('ls -lia');
    });

    it('should recompose a concise command without extra spaces', async () => {
      const command = 'ls     -l   -i    -a';
      await exec(command);
      expect(run.mock.calls[0][0]).toEqual('ls -l -i -a');
    });
  });

  describe('options', () => {
    describe('aliases', () => {
      const command = 'tree';

      beforeEach(() => {
        run.mockResolvedValue();
      });

      [[], true, null, 1].forEach(aliases => {
        // prettier-ignore
        it(`should throw an error because "${JSON.stringify(aliases)}" is not an object`, () => {
          const fn = () => exec(command, undefined, { aliases });
          expect(fn).toThrowErrorMatchingSnapshot();
        });
      });

      it('should be able to map a single alias', async () => {
        const args = { level: 1 };
        const aliases = { level: 'L' };

        await exec(command, args);
        expect(run.mock.calls[0][0]).toEqual('tree --level=1');

        await exec(command, args, { aliases });
        expect(run.mock.calls[1][0]).toEqual('tree -L 1');
      });

      it('should be able to map multiple aliases', async () => {
        const args = { level: 1, user: true, group: true, size: true };
        const aliases = { level: 'L', user: 'u', group: 'g', size: 's' };

        await exec(command, args);
        expect(run.mock.calls[0][0]).toEqual('tree --level=1 --user --group --size');

        await exec(command, args, { aliases });
        expect(run.mock.calls[1][0]).toEqual('tree -L 1 -u -g -s');
      });
    });

    describe('errorIgnored', () => {
      const command = 'non-existent-command';

      beforeEach(() => {
        run.mockImplementation(require.requireActual('runjs').run);
      });

      [[], null, 1].forEach(errorIgnored => {
        // prettier-ignore
        it(`should throw an error because "${JSON.stringify(errorIgnored)}" is not a boolean`, () => {
          const fn = () => exec(command, undefined, { errorIgnored });
          expect(fn).toThrowErrorMatchingSnapshot();
        });
      });

      it('should be able to reject an error by default', () => {
        const promise = exec(command);
        return expect(promise).rejects.toThrowErrorMatchingSnapshot();
      });

      it('should swallow the error when passing a true value', () => {
        const promise = exec(command, undefined, { errorIgnored: true });
        return expect(promise).resolves.toBeInstanceOf(Error);
      });
    });
  });
});
