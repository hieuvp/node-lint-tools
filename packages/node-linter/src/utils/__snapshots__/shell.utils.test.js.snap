// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exec command should throw an error because "     " is an empty string 1`] = `"Command \\"     \\" cannot be empty"`;

exports[`exec command should throw an error because " " is an empty string 1`] = `"Command \\" \\" cannot be empty"`;

exports[`exec command should throw an error because "" is an empty string 1`] = `"Command \\"\\" cannot be empty"`;

exports[`exec command should throw an error because "[]" is not a string 1`] = `"Expected \\"String\\", instead got \\": object\\""`;

exports[`exec command should throw an error because "1" is not a string 1`] = `"Expected \\"String\\", instead got \\"1: number\\""`;

exports[`exec command should throw an error because "null" is not a string 1`] = `"Expected \\"String\\", instead got \\"null: object\\""`;

exports[`exec command should throw an error because "true" is not a string 1`] = `"Expected \\"String\\", instead got \\"true: boolean\\""`;

exports[`exec command should throw an error because "undefined" is not a string 1`] = `"Expected \\"String\\", instead got \\"undefined: undefined\\""`;

exports[`exec options errorIgnored should throw because ""string"" is not a boolean 1`] = `"Expected \\"Boolean\\", instead got \\"string: string\\""`;

exports[`exec options errorIgnored should throw because "[]" is not a boolean 1`] = `"Expected \\"Boolean\\", instead got \\": object\\""`;

exports[`exec options errorIgnored should throw because "1" is not a boolean 1`] = `"Expected \\"Boolean\\", instead got \\"1: number\\""`;

exports[`exec options errorIgnored should throw because "null" is not a boolean 1`] = `"Expected \\"Boolean\\", instead got \\"null: object\\""`;
